### ISSUE TITLE
Figure size in RStudio is different than in rendered output

### ISSUE THREAD

**Issue by slowkow** [rstudio/rstudio#4521]
### System details

	RStudio Edition : Desktop
	RStudio Version : 1.3.74
	OS Version      : macOS 10.13.6
	R Version       : 3.5.1

### Steps to reproduce the problem

Create a new `.Rmd` file with this content:

````
---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r mtcars, fig.width = 2, fig.height = 1}
library(ggplot2)

ggplot(mtcars, aes(wt, mpg)) +
  geom_point()
```
````

Run the chunk to see the output inside RStudio.

Knit the document to see the output in the rendered HTML.

### Describe the problem in detail

The size of the figure in the RStudio interface is different than the size in the rendered output.

I spent time trying to adjust the `fig.width` and `fig.height` so the figure looks good when I run the chunk in RStudio.

It turns out my time is wasted because the rendered output does not use the same figure size as the figured in RStudio (see screenshot below).

### Questions 

- How are users supposed to size their figures properly?

- Are we meant to re-knit the file each time we tweak the `fig.width` and `fig.height` parameters?

  - If so, notice that the user must ignore the output as shown in RStudio while tweaking chunk options, since they will never share RStudio output with others and it is not representative of the knitted output.

![image](https://user-images.githubusercontent.com/209714/55044514-fd3f6580-5010-11e9-969d-25db07b002eb.png)

**Comment by david-cortes** [comment]
The size of the plots in RStudio will also change if you resize the panel, if you click zoom, etc. How else would you expect it to work? you're not setting a pre-defined size in terms of pixels for your plot in there.

**Comment by slowkow** [comment]
@david-cortes No, it does not change when I resize the panel. See the gif below.

![ezgif com-optimize (6)](https://user-images.githubusercontent.com/209714/55080040-c8b0c580-5073-11e9-9846-7e31d0a979ba.gif)

**Comment by slowkow** [comment]
### RStudio ignores dpi, but respects fig.width and fig.height

It seems that RStudio ignores the `dpi` option, but it respects the `fig.width` and `fig.height` options. This is not expected behavior.

If RStudio respects `fig.width` and `fig.height`, then I also expect it to respect `dpi`.

![image](https://user-images.githubusercontent.com/209714/55086925-9c02ab00-507f-11e9-8017-ee1dfc4e11a7.png)

### The two output devices are different

After trying a few more examples, I cannot get the output in RStudio to look the same as the output in the rendered document.

I guess the graphics devices and their settings are different for the two outputs. I would have expected them to be more similar.

This leaves the user in an undesirable situation:

- The user cannot predict the appearance of the rendered output without knitting the document.

- RStudio misleads the user to think the rendered output will look like the output shown inside the RStudio interface. Then the user will be surprised when the output does not match what they saw while writing code and tweaking the chunk options.

- Confusingly, the user must remember that a poor output in the RStudio interface is actually *desirable*. This is because some chunk options are ignored in the RStudio interface (e.g. dpi) and because the user certainly cares more about the knitted output that will be shared with others, not about the RStudio interface that will never be seen by others.

**Comment by ronblum** [comment]
@slowkow Thank you for reporting this, especially describing your experience in such detail. As you know, RStudio 1.3 is in early stages of development and is meant for testing purposes only, but thank you for trying it out and letting us know issues that you run into!

**Comment by ronblum** [comment]
See #4913 for a detailed example of the same behavior.

Note: this is the behavior in 1.1 and 1.2, so it's not new.

**Comment by alyamahmoud** [comment]
when passing fig.asp  as in {r chunk_name, echo=TRUE, fig.asp = 1, fig.height=x, fig.width=y}, the issue was solved.

**Comment by charlesli37** [comment]
> when passing fig.asp as in {r chunk_name, echo=TRUE, fig.asp = 1, fig.height=x, fig.width=y}, the issue was solved.

worked for me as well, after trying different values for fig.asp and fig.width 

> fig.asp: (NULL; numeric) The aspect ratio of the plot, i.e., the ratio of height/width. When fig.asp is specified, the height of a plot (the chunk option fig.height) is calculated from fig.width * fig.asp.

[source](https://yihui.org/knitr/options/)

**Comment by mikebessuille** [comment]
Workaround seems good, so closing this.

**Comment by slowkow** [comment]
Has anyone shown an example where `fig.asp` resolves this issue?

I tried it just now, and I am not finding that adding `fig.asp` resolves this issue. Even with `fig.asp`, the display of the figure inside the RStudio GUI is different than the display in the rendered output.

<img src="https://user-images.githubusercontent.com/209714/107815153-f8891500-6d40-11eb-88ac-36927394e823.png" width="50%">


<img src="https://user-images.githubusercontent.com/209714/107815177-ff178c80-6d40-11eb-8f89-99afb99fa8df.png" width="50%">

**Comment by slowkow** [comment]
@mikebessuille Could I ask if you might consider re-opening this issue?

Right now, I have to intentionally choose chunk options that look bad in the RStudio GUI, because those are the options that will make the figure look good in the rendered output.

Please let me know if something is unclear. I tried my best to describe the issue as clearly as I could, but I'd be happy to show you more screenshots or GIFs if that might help.

I'm currently using RStudio Version 1.4.1103 and the issue seems to be identical to what it was when I first reported it.

**Comment by jmcphers** [comment]
R Notebooks ignore a variety of chunk settings when rendering plots because they target the editor canvas for rendering (and in most cases will re-render themselves to fit the canvas). I'll re-open for consideration since we could in fact respect `dpi` here, but we'll never be able to respect all the options w/o introducing some awkward behavior in notebook mode.

https://github.com/rstudio/rstudio/blob/5ceee580ed6f632cdcc68f5978545e0b23a5a704/src/cpp/session/modules/rmarkdown/NotebookExec.cpp#L163-L176

**Comment by ctriandafillou** [comment]
I strongly agree that this is an issue worth fixing (especially since it only seems to affect Retina displays?) -- when I render a document on a non-Retina display (here my external monitor) the image that appears inline and in the output document are identical (the desired behavior):

![Screen Shot 2021-05-13 at 12 03 36 PM](https://user-images.githubusercontent.com/12089131/118174372-93969700-b3e3-11eb-9797-2a710d991ec2.png)


If I move the RStudio window to my other display (the laptop screen with a Retina display) the inline plot still looks fine

<img width="1262" alt="Screen Shot 2021-05-13 at 12 04 02 PM" src="https://user-images.githubusercontent.com/12089131/118174382-95605a80-b3e3-11eb-97db-35cd4f89789d.png">

However when I re-run the chunk the output plot is different than the rendered output:

<img width="1383" alt="Screen Shot 2021-05-13 at 12 04 56 PM" src="https://user-images.githubusercontent.com/12089131/118174384-95f8f100-b3e3-11eb-9a75-25313b1d63bc.png">

This is with RStudio Version 1.4.1106; R Version 4.0.5 (2021-03-31). The external monitor is a Dell U2412M with 1920x1200@60Hz resolution; the laptop is a 2021 MacBook Pro with macOS 11.2.3


I want to second what slowkow said about this affecting basic usage -- if the inline plots don't reflect the final output, you have to re-knit every time you want to change any plot, a slow and painful process. This seems to not be an issue with other displays and dramatically affects usability and quality.

**Comment by EoinTravers** [comment]
> when passing fig.asp as in {r chunk_name, echo=TRUE, fig.asp = 1, fig.height=x, fig.width=y}, the issue was solved.

This has no effect for me: the aspect of the image is changed, but the image is still much larger on the retina display.

Echoing the points above, this is a huge day-to-day issue for me. I don't know if this problem is affecting all MacBook useres?

**Comment by kevinushey** [comment]
If I understand correctly, the heart of the issue here is that RStudio's inline notebook images don't respect the DPI of the current display / plot being rendered, leading to the proportions of the generated plot to be inconsistent with what one would see in a "full" render.

**Comment by EoinTravers** [comment]
Playing about a bit, I can see that figures on the retina display are twice the size the should be.
External monitor:
![image](https://user-images.githubusercontent.com/3819320/123401010-4428b680-d59e-11eb-84d1-994c566c794c.png)

Retina (with `fig.width` and `fig.height`) adjusted:

![image](https://user-images.githubusercontent.com/3819320/123401130-6589a280-d59e-11eb-8bc3-a1300acb7689.png)

Retina with  ```{r fig.width=4, fig.height=4}

![image](https://user-images.githubusercontent.com/3819320/123401440-b26d7900-d59e-11eb-965e-424d633972b8.png)



I guess adding `*0.5` to all code chunk options would be a workaround, but it's not one I love.

**Comment by mjskay** [comment]
Just adding on since everything mentioned so far is on MacOS: this problem also affects high-dpi displays on Windows. On Windows you can work around this somewhat by disabling high DPI support for the RStudio application, though it is not a great workaround (since then the RStudio application becomes blurry looking).

**Comment by arlandan914** [comment]
I am having the similar issue and it's really annoying. I am using mac with an extended monitor. It worked for me this way - Go to the display settings and adjust the resolution, just switch it to higher or lower...the plot in the Rmarkdown file changes accordingly...

**Comment by netique** [comment]
This issue does not affect only retina or high-DPI displays. See my issue here for more details: https://github.com/rstudio/rmarkdown/issues/2262

Note that if you specify any chunk option related to fig, the preview is hardly readable and is way too inconsistent with the result rendered in, say, PDF. As the others mentioned, to see any changes, you have to render the document from scratch every time you change something (you can alleviate it to some extent by caching the chunks, of course).

**Comment by kevinushey** [comment]
This is where RStudio tries to handle these figure settings:

https://github.com/rstudio/rstudio/blob/3136f2997530d9318141177247b856d451ee04bc/src/cpp/session/modules/rmarkdown/NotebookExec.cpp#L169-L206

Presumedly, we need to adjust these depending on (the plot DPI? the monitor DPI? `fig.retina`?)

See also:

https://github.com/yihui/knitr/blob/83fb5084daa1161d3ee2f000b637e48bdcf64617/R/utils.R#L293-L301

**Comment by netique** [comment]
Note there is also https://github.com/rstudio/rstudio/blob/6e803f1628196c4d86f939dd971d7887620f2d19/src/cpp/session/modules/NotebookPlots.R involved in the process of inline plot creation.

**Comment by netique** [comment]
I am not 100% sure this is about the DPI. Interestingly, `grDevices::dev.size()` gives way too large dimensions when `fig.width` (or any dimension related) chunk option is provided:

<img width="358" alt="Snímek obrazovky 2022-08-19 v 15 26 43" src="https://user-images.githubusercontent.com/34926852/185628850-bf12e174-9a1f-4549-a528-f1fab7deaa69.png">

Furthermore, the first chunk device size changes with the source pane dimensions, but the second one does not (this is OK I think).

In the knitted HTML document, the dimensions are correct, of course, as the plots are generally OK as well:
<img width="235" alt="Snímek obrazovky 2022-08-19 v 15 00 05" src="https://user-images.githubusercontent.com/34926852/185623724-b4f161bc-78f9-4358-8a76-1feb3c4c3cd9.png">

[This gist ](https://gist.github.com/debruine/baa61e561935d5e0df1ea0f294399c6c)provides a wonderful hack and it is all around dimensions.

I am hopeless navigating RStudio source, but when I `debug(.rs.createNotebookGraphicsDevice)`, I can see that for the first chunk, the values are in pixels already, but the second one are in inches. This is somehow converted in https://github.com/rstudio/rstudio/blob/dad40bb763a777cffccc321de48d2d6e5dc39fa1/src/cpp/session/modules/rmarkdown/NotebookPlots.cpp#L410-L412 but I do not understand that at all...

**Comment by netique** [comment]
In `.rs.createNotebookGraphicsDevice`, when I change `pixelRatio = 1` and set res to some pleasant value such as 300, I get output with exactly the same properties as in the output document (except maybe the DPI, but the plot element scaling is top-notch which is absolutely crucial).

**Comment by netique** [comment]
@ronblum https://github.com/rstudio/rstudio/pull/12324 is a PR moved to https://github.com/rstudio/rstudio/pull/12350. Have a look at the first one where I try to enumerate issues that are duplicates to this. This is to be solve after a merge in Cherry Blossom.

**Comment by andrewheiss** [comment]
Using today's Cherry Blossom 2023.03.0-daily+3, https://github.com/rstudio/rstudio/pull/12324 fixes this for me!

![Artboard 1@3x](https://user-images.githubusercontent.com/73663/205320939-252c275b-fdab-4443-bb91-04a6730e81c0.png)

![Artboard 1 copy@3x](https://user-images.githubusercontent.com/73663/205320982-1838be92-0347-41e2-9861-1b7cc93d5bb2.png)

**Comment by netique** [comment]
@andrewheiss Can you (or anybody) please try it using different setting or display resolution? It was fixed by my PR https://github.com/rstudio/rstudio/pull/12324, so I want to be sure it works as intended and no bugs were introduced ;-)

**Comment by andrewheiss** [comment]
@netique It works correctly at all the different resolution scaling settings that macOS offers here. The new macOS Ventura settings window hides the actual dimensions, but they're 1024x665, 1147x745, 1352x878, 1512x982 (default) and 1800x1169 here

<img width="475" alt="image" src="https://user-images.githubusercontent.com/73663/205331572-f40331cc-6775-4417-9c5c-ab4cf14a247a.png">

**Comment by netique** [comment]
@andrewheiss Thanks! Still cannot believe that the fix was so trivial after all.

**Comment by jpquast** [comment]
I just encountered the problem of a different image size in the chunk output vs. the rendered figure as well (I am using a retina display). I updated RStudio hoping it would fix the issue as described above, but it unfortunately didn't. Previously I had way to large dimensions for the chunk output as described above by @netique. 

I use `dev.size()` in order to check what is going on. 

My chunk options are `dpi=300, fig.width=5, fig.height=3` 

When running the chunk the output of `dev.size()` is:

`[1] 10  6`

When rendering it is (as expected):

`[1] 5  3`

Do I miss some chunk option here? I am a bit confused about this issue still persisting. 
Would be great if someone has an idea!

**Comment by jpquast** [comment]
A thing I just noticed that seems even more concerning is that the plot in the chunk output does not seem to even have the dimensions `[1] 10  6`. Even though this is the output of `dev.size()` the figure is larger. When I resize the "Plot" tab on the right site of RStudio to have the dimension 10/6 (based on `dev.size()`), the resulting plot is smaller than the one returned by the chunk with supposedly the same dimensions.

**Comment by netique** [comment]
@jpquast Can you please report the exact RStudio version that you use? Also, can you provide any screenshots or the plots themselves (in the chunk inline preview, you can right-click on the plot and save)? Is it about plot elements scaling, resolution or margins?

**Comment by jpquast** [comment]
Thanks for the quick answer! It was indeed my mistake. I thought that the changes must already be in the release version since this was posted last month. I now installed the daily build (2023.03.0 Build 41) and it fixed the first issue that I was describing. When I use `dev.size()` the chunk output is now identical with the rendered plot in the .html.

It however still seems that there is a difference between the "Plot" tab and the chunk output.

<img width="1710" alt="Screenshot 2023-01-11 at 00 01 05" src="https://user-images.githubusercontent.com/59307482/211680284-c9163d7b-4a84-44ee-aceb-8bd041bbc246.png">

As you can see, the output of `dev.size()` is `[1] 10.069444  5.944444` when run in the console. This is the same size as I set for the chunk options. The resulting plots are however different.

Do I miss something here? Are there any specific scalings that are happening under the hood for these two different ways to return the plots? 

Thanks a lot for the help!

**Comment by jonvanausdeln** [comment]
Tested against latest daily 2023.03.0-daily+280 on Windows 11.

Original issue of the rendered (knitted) output being displayed differently still exists. In particular, when using the `dpi=` options.  I don't think this is urgent for Cherry Blossom.

Moving back to triage for prioritization.

**Comment by dcaud** [comment]
This issue adds hours to figuring out how to make plots look as expected. It feels urgent to me.

**Comment by netique** [comment]
@nikita-telkar Have you tried any recent daily releases? I think this should be fixed even in the latest stable. Could you take a try?

**Comment by andrewheiss** [comment]
Revisiting this really quick. Everything works great and as expected when using inline chunk options like `fig.width` and `fig.height`, but RStudio fails to recognize the dimensions when using YAML-style chunk options:

````qmd
```{r fig.width=7, fig.height=2}
# Correct dimensions
plot(cars)
```

```{r}
#| fig-width: 7
#| fig-height: 2
# Dimensions ignored
plot(cars)
```
````

<img width="914" alt="image" src="https://github.com/rstudio/rstudio/assets/73663/756c963b-ded4-4a3d-acc4-70576bf0bcfd">

**Comment by netique** [comment]
@andrewheiss That's already tracked in #12305

**Comment by jgutman** [comment]
@andrewheiss @netique is this issue still relevant then? From reviewing the history and linked issues it seems like this may have been solved by the fixes for #11708 (duplicated by #12305) and #12350/#12324

**Comment by andrewheiss** [comment]
Oh cool, yeah, this issue isn't relevant—the resolution issues were fixed earlier, and #11708 addresses the YAML-style dimensions
